Wrapping up GEMINI development

haitamlaarabi opened the pull request:
+ Integrating Advanced Public SPMC
+ Integrating Ridehail SPMC
+ Updating ScaleUpCharging
+ Updating HELICS deployment
+ Refactoring

haitamlaarabi committed:
update script

haitamlaarabi committed:
updating jupyter notebook

haitamlaarabi committed:
adding ridehail spmc

haitamlaarabi committed:
adding tazId and siteId to returned messages

haitamlaarabi committed:
fixing python import in cosimulation

haitamlaarabi committed:
update julius scripts

haitamlaarabi committed:
updaste scripts

haitamlaarabi committed:
adding files

haitamlaarabi committed:
update scripts

haitamlaarabi committed:
Merge remote-tracking branch 'origin/develop' into hl/integrating-ridehail-spmc

haitamlaarabi committed:
change spmc_v4.txt to py

haitamlaarabi committed:
bringing change from hl/gemini-develop

haitamlaarabi committed:
updating scripts

haitamlaarabi commented:
Test Now!

haitamlaarabi committed:
fixing PowerManagerSpec

haitamlaarabi commented:
Test now!

haitamlaarabi committed:
updating site power manager

JustinPihony commented on aws/src/main/python/beam_lambda/lambda_function.py:
```diff
@@ -105,6 +104,28 @@
+            echo "recorder started"
+            sleep 5s
+            cd -
+      path: /home/ubuntu/install-and-run-helics-scripts.sh
```
Does anything call this?

JustinPihony commented on src/main/resources/beam-template.conf:
```diff
@@ -391,7 +391,7 @@ beam.agentsim.chargingNetworkManager {
     beamFederatePrefix = "String | BEAM_FED"
     beamFederatePublication = "String | CHARGING_VEHICLES"
-    spmFederatePrefix = "String | SPM_FED_TAZ"
+    spmFederatePrefix = "String | SPM_FED"
```
You need to regenerate the code - I do not see that checked in.

haitamlaarabi replied:
Done

JustinPihony commented on src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala:
```diff
@@ -49,11 +50,11 @@ class ChargingFunctions(
-        inquiry.reservedFor.managerType == VehicleManager.TypeEnum.RideHail || inquiry.beamVehicle
-          .exists(v => v.isRideHail && (inquiry.parkingDuration <= 3600 || v.isCAV))
+        inquiry.parkingDuration <= 3600 && (inquiry.reservedFor.managerType == VehicleManager.TypeEnum.RideHail || inquiry.beamVehicle
+          .exists(_.isRideHail))
```
This lost the CAV check - was that on purpose?

haitamlaarabi replied:
I have updated the whole section

JustinPihony reviewed (changes requested):
I was light on the R and Python review. Requesting change because of the missing BeamConfig file

haitamlaarabi committed:
fixing spmc bugs

haitamlaarabi committed:
vehicle.stall.get bug

haitamlaarabi committed:
update scripts

haitamlaarabi committed:
adding spmc_v5

haitamlaarabi committed:
integrating spmc_v5,but still failing

haitamlaarabi committed:
fixing spmc integration

haitamlaarabi committed:
adding ESS info to infrastructure file

haitamlaarabi committed:
renaming loadManagement to sitePowerManager

HoneyTauOverTwo commented on src/main/python/gemini/cosimulation/nrel_spmc_controller/rudimentary_spmc_v5.py:
```diff
@@ -0,0 +1,85 @@
+#=========================================================     
+class SPM_Control():
+    def __init__(self, time_step_mins, max_power_evse, min_power_evse):
+        self.time_step_mins = 1                   # time step: 1 minutes
```
This is ignoring the parameter time_step_mins. Is this intentional?
(just saw spmc_v3.py version of this class uses it, this looks intentional)

haitamlaarabi replied:
fixed

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo commented on src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala:
```diff
@@ -119,7 +125,6 @@ object ParkingZone extends LazyLogging {
       geoId,
       parkingType,
       VehicleManager.AnyManager,
-      Some(SitePowerManager.createId(DefaultParkingZoneId.toString)),
```
PowerManagerSpec.scala#L45, SitePowerManager.scala#L95 and ZonalParkingManagerSpec.scala#L624 have to be modified to reflect this change.
(the main change is on line 132 actually)

haitamlaarabi replied:
Done

HoneyTauOverTwo reviewed (changes requested):
As pointed out by @JustinPihony, BeamConfig.scala needs to be generated again since beam-template.conf was modified.

Also, there are some tests that need updating since the field ``siteIdMaybe`` was removed from ``ParkingZone.scala``

nikolayilyin commented:
test!

haitamlaarabi committed:
updating cost scenarios

haitamlaarabi committed:
fixing hasValidChargingCapability

haitamlaarabi committed:
fixing ifRideHailCurrentlyOnShiftThenFastChargingOnly

haitamlaarabi committed:
addressing reviews from @HoneyTauOverTwo

haitamlaarabi commented:
Test now!

haitamlaarabi committed:
fixing failing tests in ElectricVehicleChargingBehaviorTest

haitamlaarabi commented:
Test now!

HoneyTauOverTwo commented on src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala:
```diff
@@ -0,0 +1,566 @@
+      ("type", a => a.equals("ChargingPlugInEvent"))
+    ).map(e => e.getAttributes.get("vehicle")).distinct
+
+    vehiclesCharged.size should be > 0 withClue ", every single vehicle should had charged at least once."
```
Originally

`vehiclesCharged.size shouldEqual 50 withClue ", every single vehicle should had charged at least once."`

I guess this test was failing to you and you changed it to `> 0`, is that it?
In this case, the clue text should be changed to match the test.

I also might have to investigate why the original check is not passing anymore, originally, I tried to make sure these tests were not flaky.

HoneyTauOverTwo commented on src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala:
```diff
@@ -0,0 +1,566 @@
+
+    val centerTAZs = List("10", "11")
+    val borderTAZs = List("20", "21", "30", "31")
+
```
~~There used to be a line like this here and other places:~~

~~`val unsuitableChargersTAZs = List("18", "19", "28", "29", "38", "39")`~~

~~which was used to check whether vehicles were plugging into chargers without sufficient power rating. What was the reason for removing this? Was it failing or this just does not make sense to check?~~

~~I have added this check to cover a change I made to how charging stalls are filtered: https://github.com/LBNL-UCB-STI/beam/pull/3727/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR126~~

I just saw you changed how that filter works and now it does not depend on charger power, just it's capability. So the test changes related to this make sense.

HoneyTauOverTwo reviewed (changes requested):
All these files:

`test/test-resources/ElectricVehicleChargingBehaviorTestData/...`
`    taz-centers.csv`
`    taz-parking-destination-only.csv`
`    taz-parking-enroute-only-free.csv`
`    taz-parking-enroute-only-mixed-prices.csv`
`    taz-parking-ride-hail.csv`
`    taz-parking-ride-hail-charging-time-cost.csv`
`    taz-parking-ride-hail-driving-time-cost.csv`

Should remove every single entry that has the TAZs `18, 19, 28, 29, 38, 39` since they are not used anymore on `ElectricVehicleChargingBehaviorTest.scala` and since `ChargingFunctions.scala` was also changed, now these chargers are actually valid, which I think is the reason why `ElectricVehicleChargingBehaviorTest` is failing:

`27 was not greater than 41, agents more likely to enroute charge in center TAZs than in border ones.`

Line 120 on `ElectricVehicleChargingBehaviorTest.scala` should also change to reflect these changes (I couldn't add a comment directly there):

```
    TAZ MAP:
                     ---------- 20, 22 ------------ 21, 23 ----------
                     |      167692.88,2213.6    168807.15,2213.6    |
             home(8) ---------- 10, 12 ------------ 11, 13 -- -------- work(9)
    166021.46,1106.8 |      167692.88,1106.8    168807.15,1106.8    | 169921.40,1106.8
                     ---------- 30, 32 ------------ 31, 33 ----------
                            167692.88,5.53      168807.15,5.53
```

haitamlaarabi committed:
updating files used in tests

haitamlaarabi commented:
Test now!

haitamlaarabi commented:
> All these files:
> 
> `test/test-resources/ElectricVehicleChargingBehaviorTestData/...` ` taz-centers.csv` ` taz-parking-destination-only.csv` ` taz-parking-enroute-only-free.csv` ` taz-parking-enroute-only-mixed-prices.csv` ` taz-parking-ride-hail.csv` ` taz-parking-ride-hail-charging-time-cost.csv` ` taz-parking-ride-hail-driving-time-cost.csv`
> 
> Should remove every single entry that has the TAZs `18, 19, 28, 29, 38, 39` since they are not used anymore on `ElectricVehicleChargingBehaviorTest.scala` and since `ChargingFunctions.scala` was also changed, now these chargers are actually valid, which I think is the reason why `ElectricVehicleChargingBehaviorTest` is failing:
> 
> `27 was not greater than 41, agents more likely to enroute charge in center TAZs than in border ones.`
> 
> Line 120 on `ElectricVehicleChargingBehaviorTest.scala` should also change to reflect these changes (I couldn't add a comment directly there):
> 
> ```
>     TAZ MAP:
>                      ---------- 20, 22 ------------ 21, 23 ----------
>                      |      167692.88,2213.6    168807.15,2213.6    |
>              home(8) ---------- 10, 12 ------------ 11, 13 -- -------- work(9)
>     166021.46,1106.8 |      167692.88,1106.8    168807.15,1106.8    | 169921.40,1106.8
>                      ---------- 30, 32 ------------ 31, 33 ----------
>                             167692.88,5.53      168807.15,5.53
> ```

I updated those files

HoneyTauOverTwo committed:
Fixes TAZ map comment on ElectricVehicleChargingBehaviorTest.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo reviewed (approved):
Looks good to me.

haitamlaarabi commented:
Test now!

HoneyTauOverTwo commented:
The issue came back with the test getting stuck

```
18:26:52 *** Test still running after 46 minutes, 22 seconds: suite name: ElectricVehicleChargingBehaviorTest, test name: Ride Hail Electric vehicles should pick chargers choosing smaller DrivingTimeCost.. 
18:26:56 21:26:58.839 [ClusterSystem-akka.actor.default-dispatcher-16] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
18:26:58 21:27:08.859 [ClusterSystem-akka.actor.default-dispatcher-15] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
18:27:08 21:27:18.879 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
```

We should take care of this before merging. 2 out of 3 attempts got stuck with this.

HoneyTauOverTwo commented on src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala:
```diff
@@ -412,34 +444,34 @@ class ElectricVehicleChargingBehaviorTest
     val closestTAZsCAV = List("12")
```
The test `"Ride Hail Electric vehicles" should "pick chargers choosing smaller DrivingTimeCost."` is sometimes getting stuck on CI, this should be fixed before merging. I'll investigate.

HoneyTauOverTwo commented:
test!

this one passed: 2 fails out of 4 tries

HoneyTauOverTwo commented:
test!

failed but with flaky test:
`beam.sim.BeamWarmStartRunSpec.Beam WarmStart must run beamville scenario for two iterations with warmstart with normal and fake skims`

HoneyTauOverTwo committed:
Fixes parking file for mixed prices enroute test and puts back the check to ensure every single vehicle is enrouting.

HoneyTauOverTwo commented:
test!

haitamlaarabi committed:
update tests

haitamlaarabi commented:
Test now!

HoneyTauOverTwo commented:
Latest test got stuck as well. I could not reproduce this issue locally.

I ran the test getting stuck (`Ride Hail Electric vehicles should pick chargers choosing smaller DrivingTimeCost.`) 10 times and had no issues.

I also ran the hole class tests another 7 times and did not encounter any issues (I have seen tests behave a bit different in the past when running them separatelly compared to running the hole class).

Although the tests don't get stuck, I did notice the ChargingNetwork error message does show up locally, but it is only once per test (when it does show):

`ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer`

I also notice a similar error message with more info:

`ERROR b.a.agents.ridehail.RideHailAgent - unhandled event: Event(BeamAgentSchedulerTimer,RideHailAgentData(rideHailVehicle-07-L5@GlobalRHM (RH_BEV_L5,8.785613000000001km),Vector(rideHailVehicle-07-L5@GlobalRHM),,0,List(),None)) in state [ Idle ] - vehicle( rideHailVehicle-07-L5@GlobalRHM )`

These detailed errors always refer to fully automated vehicles, but I am unsure if they are the cause for the other error message which seems to be the cause for the test getting stuck.

I will temporarily change the stuck test so it does not have any ride hail vehicle with human driver and run CI a few times to see if it still gets stuck.

HoneyTauOverTwo committed:
Changes ride hail fleet for the test getting stuck "Ride Hail Electric vehicles" should "pick chargers choosing smaller DrivingTimeCost." to contain only Level 5 automated vehicles to investigate if the source of the problem can be isolated to RH with human driver or L5.

HoneyTauOverTwo commented:
test!

This one passed and the console output ~~does not have any mention of "unhandled message"~~. (actually does, forgot to look at the full log)

HoneyTauOverTwo commented:
test!

Same as before.

JustinPihony commented:
I've questioned the changes at [this point](https://github.com/LBNL-UCB-STI/beam/pull/3729/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR49) and it seems plausible it could be causing this?

HoneyTauOverTwo commented:
> I've questioned the changes at [this point](https://github.com/LBNL-UCB-STI/beam/pull/3729/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR49) and it seems plausible it could be causing this?

You're right on the money! I discovered a bit ago that vehicles are trying to charge on parking spots with no charging capability. 
The changes made to `hasValidChargingCapability` are causing this. I am pushing changes in a bit.

HoneyTauOverTwo committed:
Adds back TAZs 18, 19, 28, 29, 38 and 39 were actually still needed for a test.

HoneyTauOverTwo committed:
Merge branch 'hl/integrating-ridehail-spmc' of https://github.com/LBNL-UCB-STI/beam into hl/integrating-ridehail-spmc

HoneyTauOverTwo committed:
Fixes bug introduced in ChargingFunctions.hasValidChargingCapability method modification.

Parking zones without charging capability would not have the block "zone.chargingPointType.forall" executed, and the result was true. This would make EVs that intended to recharge to park somewhere without a charger.

HoneyTauOverTwo committed:
Adds back the test checking whether every single vehicle was charging. The removal of this was hiding the issue fixed at the previous commit.

HoneyTauOverTwo commented:
@JustinPihony , @haitamlaarabi I am still unsure if the [current version of `hasValidChargingCapability`](https://github.com/LBNL-UCB-STI/beam/blob/f39d1c08b2e21cf6fd34400e6d276cc7ac3c25ee/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L124) is correct. The current one seems to imply that an EV cannot, in any circumstance, park somewhere where there is not a charger.

[The version present on develop branch](https://github.com/LBNL-UCB-STI/beam/blob/develop/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L126) will only filter out a parking zone if the inquiry is related to an activity that requires charging.

I suggest adding the `verifyCharger` check back into it if we are to allow EVs to park somewhere without a charger:

```
    private def hasValidChargingCapability(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
    val verifyCharger = inquiry.beamVehicle.isDefined &&
      inquiry.beamVehicle.get.beamVehicleType.chargingCapability.isDefined && (
        inquiry.searchMode == ParkingSearchMode.EnRouteCharging ||
        inquiry.parkingActivityType == Charge ||
        inquiry.parkingActivityType == EnRoute
      )
    if (!verifyCharger) {
      return true
    }
    zone.chargingPointType match {
      case Some(chargingPointType) =>
        inquiry.beamVehicle.forall {
          case beamVehicle if beamVehicle.isEV =>
            val vehicleHasDCFasChargingCapability =
              beamVehicle.beamVehicleType.chargingCapability.forall(ChargingPointType.isFastCharger)
            val stationHasDCFasChargingCapability = ChargingPointType.isFastCharger(chargingPointType)
            val hasInvalidChargingCapability = !vehicleHasDCFasChargingCapability && stationHasDCFasChargingCapability
            !hasInvalidChargingCapability
          case _ => false
        }
      case _ =>
        false
    }
  }
```

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo committed:
Fixes temporary version of "rideHailFleet0.7soc.csv" that was accidentally pushed.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo committed:
Fixes parking filters returning true when zone.chargingPointType equals None.

HoneyTauOverTwo commented:
I just found out that a similar issue exists on other filters as well. `ifRideHailCurrentlyOnShiftThenFastChargingOnly` for instance:

![image](https://github.com/LBNL-UCB-STI/beam/assets/31476381/573ddbdd-7217-4ffc-b7e3-a138bb84774c)

If the method consists only of this kind of statement `zone.chargingPointType.forall` when `chargingPointType` is `None`, the block inside will not be executed at all, and the result will be treated as `true`, which is the opposite of what we want in this case.

I have pushed changes that instead of using `zone.chargingPointType.forall(chargingPointType => [...] ChargingPointType.isFastCharger(chargingPointType)` uses `zone.chargingPointType.exists(ChargingPointType.isFastCharger)` which returns false when the `Option` is `None`.

HoneyTauOverTwo commented:
test!

failed with `beam.integration.EnrouteChargingSpec.Running a single person car-only scenario should do enroute upon not enough charging`. [EnrouteChargingSpec#L147](https://github.com/LBNL-UCB-STI/beam/blob/hl/integrating-ridehail-spmc/src/test/scala/beam/integration/EnrouteChargingSpec.scala#L147)

It is working locally.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo commented:
Previous CI build got stuck on the same test `"Ride Hail Electric vehicles" should "pick chargers choosing smaller DrivingTimeCost."`

It appears like the fixes made to prevent RH vehicles to try to charge on parking zones with no charger was not the cause of this issue.

HoneyTauOverTwo committed:
Puts back checking whether every vehicle is enrouting.

HoneyTauOverTwo committed:
Temporarily ignoring test getting suck to observe other CI test results.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo commented:
> @JustinPihony , @haitamlaarabi I am still unsure if the [current version of `hasValidChargingCapability`](https://github.com/LBNL-UCB-STI/beam/blob/f39d1c08b2e21cf6fd34400e6d276cc7ac3c25ee/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L124) is correct. The current one seems to imply that an EV cannot, in any circumstance, park somewhere where there is not a charger.
> 
> [The version present on develop branch](https://github.com/LBNL-UCB-STI/beam/blob/develop/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L126) will only filter out a parking zone if the inquiry is related to an activity that requires charging.
> 
> I suggest adding the `verifyCharger` check back into it if we are to allow EVs to park somewhere without a charger:
> 
> ```
>     private def hasValidChargingCapability(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
>     val verifyCharger = inquiry.beamVehicle.isDefined &&
>       inquiry.beamVehicle.get.beamVehicleType.chargingCapability.isDefined && (
>         inquiry.searchMode == ParkingSearchMode.EnRouteCharging ||
>         inquiry.parkingActivityType == Charge ||
>         inquiry.parkingActivityType == EnRoute
>       )
>     if (!verifyCharger) {
>       return true
>     }
>     zone.chargingPointType match {
>       case Some(chargingPointType) =>
>         inquiry.beamVehicle.forall {
>           case beamVehicle if beamVehicle.isEV =>
>             val vehicleHasDCFasChargingCapability =
>               beamVehicle.beamVehicleType.chargingCapability.forall(ChargingPointType.isFastCharger)
>             val stationHasDCFasChargingCapability = ChargingPointType.isFastCharger(chargingPointType)
>             val hasInvalidChargingCapability = !vehicleHasDCFasChargingCapability && stationHasDCFasChargingCapability
>             !hasInvalidChargingCapability
>           case _ => false
>         }
>       case _ =>
>         false
>     }
>   }
> ```

For some reason, this change fixes the failing test locally `beam.integration.EnrouteChargingSpec.Running a single person car-only scenario should do enroute upon not enough charging`

I'll push this change and run CI again. @haitamlaarabi, please review the quoted message and let me know if you agree with it.

HoneyTauOverTwo committed:
Adds back the "verifyCharger" check on "hasValidChargingCapability".

Without this check, this filter makes that an EV cannot, in any circumstance, park somewhere where there is not a charger.

HoneyTauOverTwo commented:
test!

passed (test getting stuck skipped)

HoneyTauOverTwo commented:
test!

(same as before)

HoneyTauOverTwo committed:
Enables the test getting stuck back on.

HoneyTauOverTwo commented:
test!

passed, including the test previously getting stuck. Will run CI another 2 times at least to make sure.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo committed:
Modifying DrivingTimeCost test to check if it is getting stuck due to the existence of parking zones without chargers.

HoneyTauOverTwo commented:
test!

CI wasn't working.

HoneyTauOverTwo commented:
test!

CI wasn't working.

HoneyTauOverTwo commented:
test!

This passed, testing again to see it if was a fluke.

HoneyTauOverTwo commented:
test!

Failed with test stuck.

HoneyTauOverTwo committed:
Removes RH vehicles with human drivers from the test getting stuck to investigate issue.

Also, reverts the parking file changed on previous commit "Modifying DrivingTimeCost test to check if it is getting stuck due to the existence of parking zones without chargers."

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
test!

Still gets stuck without RH with human drivers, so it is most likely something related to L5 RH or both.

HoneyTauOverTwo committed:
Adds back RH vehicles with human drivers and removes L5 automated vehicles from the test getting stuck to investigate issue.

HoneyTauOverTwo commented:
test!

fmt failed.

HoneyTauOverTwo committed:
fmt

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
test!

Passed.

HoneyTauOverTwo commented:
I was able to reproduce the stuck issue locally as well, and it is definitely related to parallelism.

Also, from the CI results, it seems like the issue is only related to L5 automated RH electric vehicles, so RH electric vehicles with human drivers is not the source of the issue.

HoneyTauOverTwo committed:
Adds ignore tag to AV RH test getting stuck since issue will be investigated on another PR (was not originated here)

Also, changes ride hail fleet and the related test to their original state with both automated and non-automated vehicles.

HoneyTauOverTwo commented:
test!

HoneyTauOverTwo reviewed (approved):
Approving since AV RH issue was proven to not be caused by these changes.
This issue will continue to be investigated on #3766.
